core/src/main/java/org/apache/calcite/util/format/FormatElementEnum.java

```diff
@@ -39,188 +39,207 @@
+    }
+    @Override public void format(StringBuilder sb, Date date) {
       final Calendar calendar = Work.get().calendar;
       calendar.setTime(date);
-      return String.format(Locale.ROOT, "%02d", calendar.get(Calendar.SECOND));
+      sb.append(String.format(Locale.ROOT, "%d", (calendar.get(Calendar.MONTH) / 3) + 1));
     }
   },
-  MS("The millisecond as a decimal number (000-999)") {
-    @Override public String format(Date date) {
+  MS("SSS", "The millisecond as a decimal number (000-999)") {
+    @Override public void format(StringBuilder sb, Date date) {
       final Calendar calendar = Work.get().calendar;
       calendar.setTime(date);
-      return String.format(Locale.ROOT, "%03d", calendar.get(Calendar.MILLISECOND));
+      sb.append(String.format(Locale.ROOT, "%03d", calendar.get(Calendar.MILLISECOND)));
     }
   },
-  TZR("The time zone name") {
-    @Override public String format(Date date) {
+  SS("s", "The second as a decimal number (00-60)") {
+    @Override public void format(StringBuilder sb, Date date) {
+      final Calendar calendar = Work.get().calendar;
+      calendar.setTime(date);
+      sb.append(String.format(Locale.ROOT, "%02d", calendar.get(Calendar.SECOND)));
+    }
+  },
+  TZR("z", "The time zone name") {
+    @Override public void format(StringBuilder sb, Date date) {
       // TODO: how to support timezones?
-      return "";
     }
   },
-  WW("The week number of the year (Sunday as the first day of the week) as a decimal "
+  WW("w", "The week number of the year (Sunday as the first day of the week) as a decimal "
       + "number (00-53)") {
-    @Override public String format(Date date) {
+    @Override public void format(StringBuilder sb, Date date) {
       final Calendar calendar = Work.get().calendar;
       calendar.setTime(date);
       calendar.setFirstDayOfWeek(Calendar.SUNDAY);
-      return String.format(Locale.ROOT, "%02d", calendar.get(Calendar.WEEK_OF_YEAR));
+      sb.append(String.format(Locale.ROOT, "%02d", calendar.get(Calendar.WEEK_OF_YEAR)));
     }
   },
-  YY("Last 2 digits of year") {
-    @Override public String format(Date date) {
+  YY("yy", "Last 2 digits of year") {
+    @Override public void format(StringBuilder sb, Date date) {
       final Work work = Work.get();
-      return work.yyFormat.format(date);
+      sb.append(work.yyFormat.format(date));
     }
   },
-  YYYY("The year with century as a decimal number") {
-    @Override public String format(Date date) {
+  YYYY("yyyy", "The year with century as a decimal number") {
+    @Override public void format(StringBuilder sb, Date date) {
       final Calendar calendar = Work.get().calendar;
       calendar.setTime(date);
-      return String.format(Locale.ROOT, "%d", calendar.get(Calendar.YEAR));
+      sb.append(String.format(Locale.ROOT, "%d", calendar.get(Calendar.YEAR)));
     }
   };
 
   private final String description;
+  final String javaFmt;
```

Comment:
Yes, this will be a problem. `Q` (quarter as a decimal) has no Java equivalent that I am aware of. `toPattern` is overridden for this element to throw an `UnsupportedOperationException`

```
  Q("", "The quarter as a decimal number (1-4)") {
    // TODO: Allow parsing of quarters.
    @Override public void toPattern(StringBuilder sb) throws UnsupportedOperationException {
      throw new UnsupportedOperationException("Cannot convert 'Q' FormatElement to Java pattern");
    }
...
```

I updated the javadoc for `toPattern` to reflect this as well. Not ideal but perhaps we can revisit?